### ISSUE TITLE
How to test the `with_graceful_shutdown` in axum?

### ISSUE THREAD

**Issue by HaoYang670** [JosephLenton/axum-test#88]

**Comment by JosephLenton** [comment]
Hey @HaoYang670 , sorry for not being able to look into this sooner. I had started a new role which has sucked up most of my time.

I have looked into this today, and sadly I don't see a way to easily test `Serve::with_graceful_shutdown`. The issue is that once you have called `serve()` to create the server, there is no way for Axum Test to get out the address and know where to send requests.

This is still achievable, but with limitations.

# Option 1, being able to test against `Serve`

This option boils down to being able to pass in `Serve` to `TestServer::new`, help to make this straight forward, and have a way to test if the server has ended (i.e. `TestServer::is_shutdown`).

The code would look something like this:

```rust
// 0. Setup your application
let my_app = Router::new();

// 1. You would need to pick a port for the server (which can be random, or can be predefined).
let tokio_tcp_listener = axum_test::util::new_random_tokio_tcp_listener()?;
let local_address = tokio_tcp_listener.local_addr().ok().unwrap();
let server = serve(tokio_tcp_listener, my_app.into_make_service())
    .with_graceful_shutdown(my_app_shutdown_signal());

// 2. You would need to setup which port the server is using for AxumTest.
let config = TestServerConfig::builder()
    .http_transport_with_ip_port(Some(local_address.ip()), Some(local_address.port()));
let test_server = TestServer::new_with_config(server, config);

// 3. Write your tests

// 4. I would add a way to test if the server is still running
let is_running = test_server.is_shutdown();
```

# Option 2, add `with_graceful_shutdown` to the `TestServer`

With this approach the `TestServer` gains the ability to take your graceful shutdown, and handle it for you. This is much simpler to implement in Axum Test, the example code I have is much shorter, and there is less to go wrong.

The downside is you may need to make refactors to your code to get it working.

An example test would look something like ...

```rust
// 0. Setup your application
let my_app = Router::new();

// 1. You would need to setup the graceful shutdown on the `TestServer`, not your application.
let config = TestServerConfig::builder()
    .with_graceful_shutdown( my_app_shutdown_signal() );
let test_server = TestServer::new_with_config(my_app, config);

// 3. Write your tests

// 4. I would add a way to test if the server is still running
let is_running = test_server.is_shutdown();
```

The two options above are the directions I am thinking of going in. If you see this then feel free to let me know your thoughts. I may end up implementing a mix of both of the above. I will need a little time to think about it before plowing on. Feel free to post back with any thoughts, nitpicks, concerns, etc. Or if you feel I've misunderstood your issue.

**Comment by HaoYang670** [comment]
Thanks, @JosephLenton, for your kind help. I will try your suggestions and provider you feedbacks. Many thanks.

**Comment by JosephLenton** [comment]
Sorry in case I wasn't clear, but the options I posted were for things to change in Axum Test to get this closer to working. This is still on my radar to improve.

**Comment by JosephLenton** [comment]
Hey @HaoYang670 , a method I needed to solve this has now been shipped in Axum.

I am currently preparing a large update, that will include a way to test `WithGravefulShutdown`. I am hoping to have it out in the coming week. Sorry for the delay on this!

**Comment by HaoYang670** [comment]
Thanks @JosephLenton. Looking forward to your amazing work!

**Comment by JosephLenton** [comment]
@HaoYang670 this should now be resolved in [v16.0.0](https://crates.io/crates/axum-test/16.0.0), which also requires Axum v0.7.6.

Testing graceful shutdown is doable now as 1) the `TestServer` can now take a `WithGracefulShutdown` when being built, and 2) there is a new function `TestServer::is_running` to say if the server is still running or has stopped.

I have an example of a full test here, using a notify to tell the server to shutdown. Hopefully you can change it for your needs:
```rust
use axum_test::TestServer;
use axum_test::util::new_random_tokio_tcp_listener;
use axum::routing::get;
use axum::routing::IntoMakeService;
use axum::serve;
use axum::Router;
use std::time::Duration;
use tokio::sync::Notify;
use tokio::time::sleep;
use std::sync::Arc;

async fn get_ping() -> &'static str {
    "pong!"
}

#[tokio::test]
async fn it_should_panic_when_run_with_mock_http() {
    let shutdown_notification = Arc::new(Notify::new());
    let waiting_notification = shutdown_notification.clone();

    // Build an application with a route.
    let app: IntoMakeService<Router> = Router::new()
        .route("/ping", get(get_ping))
        .into_make_service();
    let port = new_random_tokio_tcp_listener().unwrap();
    let application = serve(port, app)
        .with_graceful_shutdown(async move { waiting_notification.notified().await });

    // Run the server.
    let server = TestServer::builder()
        .build(application)
        .expect("Should create test server");

    // 1. It is running.
    server.get("/ping").await.assert_status_ok();
    assert!(server.is_running());

    // 2. Notify the service so it shuts down.
    shutdown_notification.notify_one();
    sleep(Duration::from_millis(10)).await;

    // 3. You can now assert it is not running
    assert!(!server.is_running());
}
```

Let me know how it goes. Again I am sorry for how long it's taken to get this resolved. If you run into any issues, or if I am missing something, then feel free to ask for more help.

**Comment by HaoYang670** [comment]
Thank you, @JosephLenton, for your help. Overall, I think it is what I am looking for. I will try this solution as soon as I can. Thanks
